File: codesignal/sorting/2.mergeSort.py
class Solution:
    def mergeSort(self, arr):
        if len(arr) == 1:
            return arr

        middle = len(arr) // 2
        left = arr[:middle]
        right = arr[middle:]

        return self.merge(self.mergeSort(left), self.mergeSort(right))
    
    def merge(self, left, right):
        result = []
        leftIndex = 0
        rightIndex = 0

        while leftIndex < len(left) and rightIndex < len(right):
            if left[leftIndex] < right[rightIndex]:
                result.append(left[leftIndex])
                leftIndex += 1
            else:
                result.append(right[rightIndex])
                rightIndex += 1

        return result + left[leftIndex:] + right[rightIndex:]
    
print(Solution().mergeSort([1, 5, 2, 8, 3, 4, 7, 6]))

from typing import List

# quicker solution
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def merge(arr, L, M, R):
            left, right = arr[L:M+1], arr[M+1:R+1]
            i, j, k = L, 0, 0
            while j < len(left) and k < len(right):
                if left[j] <= right[k]:
                    arr[i] = left[j]
                    j += 1
                else:
                    arr[i] = right[k]
                    k += 1
                i += 1
            while j < len(left):
                nums[i] = right[k]
                k += 1
                i += 1

        def mergeSort(arr, l, r):
            if l == r:
                return arr
            m = (l + r) // 2
            mergeSort(arr, l, m)
            mergeSort(arr, m + 1, r)
            merge(arr, l, m, r)
            return arr
        
        return mergeSort(nums, 0, len(nums) - 1)

File: codesignal/sorting/2.mergeSort.cpp

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> mergeSort(vector<int> a) {
    if (a.size() == 1) return a;
    int mid = a.size() / 2;
    vector<int> left(a.begin(), a.begin() + mid);
    vector<int> right(a.begin() + mid, a.end());
    left = mergeSort(left);
    right = mergeSort(right);
    return merge(left, right);
  }
  vector<int> merge(vector<int> left, vector<int> right) {
    vector<int> result;
    while (left.size() > 0 || right.size() > 0) {
      if (left.size() > 0 && right.size() > 0) {
        if (left.front() <= right.front()) {
          result.push_back(left.front());
          left.erase(left.begin());
        } else {
          result.push_back(right.front());
          right.erase(right.begin());
        }
      } else if (left.size() > 0) {
        result.push_back(left.front());
        left.erase(left.begin());
      } else if (right.size() > 0) {
        result.push_back(right.front());
        right.erase(right.begin());
      }
    }
    return result;
  }
};

// quicker solution
class Solution {
public:
  vector<int> sortArray(vector<int>& nums) {
    mergeSort(nums, 0, nums.size() - 1);
    return nums;
  }
private:
  void merge(vector<int>& arr, int L, int M, int R) {
    vector<int> left(arr.begin() + L, arr.begin() + M + 1);
    vector<int> right(arr.begin() + M + 1, arr.begin() + R + 1);
    int i = L;
    int j = 0;
    int k = 0;
    while (j < left.size() && k < right.size()) {
      if (left[j] <= right[k]) {
        arr[i] = left[j];
        j++;
      } else {
        arr[i] = right[k];
        k++;
      }
      i++;
    }
    while (j < left.size()) {
      arr[i] = left[j];
      j++;
      i++;
    }
    while (k < right.size()) {
      arr[i] = right[k];
      k++;
      i++;
    }
  }

  void mergeSort(vector<int>& arr, int l, int r) {
    if (l < r) {
      int m = l + (r - l) / 2;
      mergeSort(arr, l, m);
      mergeSort(arr, m + 1, r);
      merge(arr, l, m, r);
    }
  }
};